### ISSUE TITLE
Illegal submissions megathread

### ISSUE THREAD

**Issue by 0hq** [openai/parameter-golf#677]
Hey all! We're seeing a lot of invalid solutions due to autoresearch and some rules confusion. Clarifying everything here.

**Comment by andrew-medrano** [comment]
@mhuen @valerio-oai @0hq Unfortunately, there are 100+ open ngram submissions at this point and more every hour which are incorrectly normalized. The agents are just reading each other's incorrect code and overwhelming the pull request backlog. There is no easy centralized way to inform all competitors, remove/fix the current submissions, and especially prevent future bad ones. 

I don't even think the real problem is about n-grams specifically. There might eventually be another technique which has a subtle scoring error and cause similar overflow problems again. I do think a much more robust submission process with more friction/guardrails is necessary to keep the grading manageable. 

My 2-cents: Something like a two-stage submission backlog might make sense. I am not sure how you could even accomplish this with github, but all of the current PRs go into the first backlog, then promotion into the screened backlog is gated by a quick screen (by organizers or even other competitors), and the screened backlog would have a significantly higher signal to noise ratio.

**Comment by valerio-oai** [comment]
Hi again everyone! I just went through and closed a bunch of non-compliant PRs, i.e. illegal TTT and GPTQ, and, of course, a lot of hashed n-gram caches, which are artifacts of a lack of renormalization and hashing (as @abaybektursun , @Eppie , and others in this thread, #886, and other places in the repo have pointed out, too -- thank you!) instead of actual compression.

No new submissions have been merged, and, given the enormous backlog of potentially illegal submissions, I only reviewed those with 3 seeds, a valid submissions.json file, and where a SOTA record result was being claimed, to cut down on the amount of time necessary for review. This also means that unless I've made a mistake, I will now be looking for new PRs to potentially merge only after #988 . 

For the future, I would strongly recommend people follow rules like the ones set forth by @NoesisGenesis above: breaking autoregressivity can be very easy, and we want to make sure that eval-time improvements are genuine reweightings of the probability distribution, and do not accidentally leak tokens they are not meant to. 

Speaking for myself only, I would also recommend people slow down a little: this competition is exciting, and advancements in ML are fun and interesting, but there's still a long way to go, and with all likelihood, the leading submission now or in a week won't be the winning one. This should allow you to explore more, take your time with experiments and ideas, and not have to rush to be the first to submit a PR. I think there are a lot of interesting train-time improvements to be made that people haven't gotten around to, and I'm sure there is also novel test-time stuff that will be interesting.

We are still considering a second leaderboard of some type, even though this one is still intact on account of these hashed n-gram caches being invalidated.

Thank you for competing, and I look forward to reviewing more submissions :)

PRs closed/reviewed (on account of wanting to do this quickly, I couldn't give out personalized closing messages, I apologize):
#846, #868, #869, #870, #881, #888, #893, #907, #720, #724, #824, #921, #745, #769, #779, #809, #850, #880, #913, #771, #859, #688, #794, #814, #828, #843, #925, #944, #988, #727, #792, #851, #945

**Comment by Robby955** [comment]
Great work on #1145. This is consistent with the partition function analysis in #1147. The unnormalized apparent gain was ~1 BPB at 1M buckets; after proper normalization the real signal is ~0.003 BPB, matching your result and the collision-free trie baseline (PR #511, -0.0013 BPB). Two independent implementations, both normalized, both landing in the same ballpark, both 300-770x smaller than the unnormalized artifact.

**Comment by MatoTeziTanka** [comment]
## Meta-comment: 6-PR cluster of the "Pre-Quant TTT = multi-epoch fine-tune on val_tokens" pattern — unified ruling request

@valerio-oai @0hq @cocohearts — bundling this for a single ruling because the same structural code pattern now appears in six separate PRs, all claiming "Pre-Quant TTT" but implementing it as an N-epoch AdamW supervised fine-tune on `val_tokens` with no per-token scoring discipline. They all share the exact structure that @valerio-oai ruled out on PR #1376 (closed 2026-04-10):

| PR | Author | Track | Epochs on val | Line references | Status |
|---|---|---|---|---|---|
| #1376 | stukenov | record | 6 | `ttt_adapt_adamw` L825, called L1453 with `val_tokens` | **CLOSED** on 2026-04-10 (this ruling) |
| #1485 | ndokutovich | record | 6 | `prequant_ttt_adapt_adamw` | **Self-closed** by author on 2026-04-10 |
| #1487 | ndokutovich | record | 10 | `prequant_ttt_adapt_adamw` L1289, called L1956 with `val_data.val_tokens` | OPEN (reopened after #1517 landed) |
| #1488 | ndokutovich | record | 10 | `prequant_ttt_adapt_adamw` L110, same pattern, combined with SLOT-24 | OPEN |
| #1489 | joshkmartinez | record | 6 | `prequant_ttt_adapt_adamw` L1024 (docstring says "ported from PR #1423" — it isn't), called with `val_tokens` | OPEN |
| #1517 | RulinShao | record | 18/22 | `ttt_adapt_adamw` L1165, docstring literally `"""AdamW TTT: fine-tune on val data BEFORE quantization"""`, called L2209 with `val_tokens` | OPEN |
| #1539 | translatingthename | record | 6 | Val TTT block L2417-2455, `for epoch in range(args.ttt_epochs)` L2426, compiled via `torch.compile` (perf wrapper, not legality vector) | OPEN |

**The structural pattern is identical across all seven PRs:**

```python
for epoch in range(N):                                # N ∈ {6, 10, 18, 22}
    for chunk_start in range(0, len(val_tokens) - seq_len - 1, step):
        local = val_tokens[chunk_start:chunk_start+seq_len+1]  # <-- val_tokens
        x = local[:-1]; y = local[1:]
        loss = base_model(x, y)                       # standard teacher-forced CE
        loss.backward(); optimizer.step()             # unconditional SGD update
# Then: eval_val(...) on the same val_tokens tensor
```

There is **no `with torch.no_grad()` scoring accumulator before the first `optimizer.step()`** in any of these implementations. Every token in the validation set is seen by the optimizer for N passes before the final `eval_val_*` / `final_int6_sliding_window` scoring pass reads the same tokens.

This is the pattern Issue #677's ruling (on multi-epoch TTT that scores only on the final pass) and Issue #402's score-first principle were opened to disallow. Several of these PRs explicitly claim "ported from PR #1423" or "ported from PR #1416" in their docstrings, but those legal reference implementations score each `(x,y)` chunk into the sliding-BPB accumulator **before** `optimizer.step()` touches the weights derived from that chunk — a different structure from the loops above.

**Request to the mod team:** could one of you post a binding ruling on this thread (or on each PR) clarifying whether the pattern above is permitted or disallowed? A single ruling will close or clear all seven at once, plus whatever appears next under the same "Pre-Quant TTT" label.

If the ruling goes against the pattern, the engineering work in each of these PRs — depth recurrence (#1487), banked Muon + depth recurrence (#1517), parallel residuals (#1489), SP8192 recipe (#1539), `torch.compile` wrapper (#1539), SLOT-24 (#1488) — all carries over cleanly to a resubmission that trains the Pre-Quant TTT pass on a held-out slice of `fineweb_train_*.bin` instead of `val_tokens`. The rulings don't touch the architecture work; they only target the input-tensor selection for the pre-quant adaptation loop.

If the ruling goes in favor of the pattern, that's also fine — I'll happily retract the community reviews I just posted on each of these threads — but right now the cluster is growing faster than individual per-PR discussions can keep up with, and the authors deserve a clear signal so they know whether to iterate or withdraw.

Individual compliance reviews for each PR are already posted on their respective threads with the specific line-number citations. This comment is just the umbrella ping.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Each per-PR audit was performed with static code review against the head SHA plus a CPU import/hyperparameter smoke test on a dedicated container (`proteus-engine`, 128 GB RAM, Triton 3.6.0). AI tooling: reviews drafted with Claude Code (Sonnet/Opus) using an internal review template; citations verified against each PR's actual code.*

**Comment by MatoTeziTanka** [comment]
## Correction to the meta-comment above

I need to correct an error in my 6-PR cluster meta-comment above. I cited **PR #1416 (erichroepke)** and **PR #1423 (aryanbhosale)** as the "legal reference implementations" that score each `(x,y)` chunk into the sliding-BPB accumulator before `optimizer.step()`. That's wrong.

**What I got wrong:** I conflated the folder-name branding ("LegalTTT", "PreQuantTTT_QKGain5") with the actual code at the current head of each of those PRs. When I re-audited #1416 and #1423 just now, both have `ttt_adapt_adamw(args, base_model, device, val_tokens, rank=0, world_size=1, log0=print)` at line 1132, called at line 2135 under `if args.ttt_enabled:` with `val_tokens` — a multi-epoch AdamW supervised fine-tune on `val_tokens` with no per-chunk score-first discipline. **That's the exact pattern this thread is asking to be ruled out**, not the legal contrast.

**The actual legal reference implementation I should have cited:**

**PR #1413 (dexhunter)** — the current leaderboard entry at **val_bpb 1.0828**, merged as *SP8192 + QK-Gain 5 + Legal Score-First TTT*. #1413's train_gpt.py is an lzma+b85 self-extracting shim; decompressed, the TTT loop reads:

```python
# per-chunk eval scoring
nll = F.cross_entropy(logits.reshape(-1, logits.size(-1)).float(),
                     y_batch.reshape(-1), reduction='none').reshape(bsz, seq_len)
for i, ws in enumerate(batch_ws):
    wlen = wlens[i]
    s = 0 if ws == 0 else context_size
    scored_nll = nll[i, s:wlen].to(torch.float64)
    loss_sum += scored_nll.sum()
    token_count += float(wlen - s)
    # ...

# then, only if not the last chunk, adapt on that chunk
is_last_chunk = ci == num_chunks - 1
if not is_last_chunk and h.ttt_epochs > 0:
    base_model.train()
    chunk_seqs = (chunk_end - chunk_start) // seq_len
    if chunk_seqs > 0:
        cos_lr = h.ttt_lr * .5 * (1. + math.cos(math.pi * ci / max(num_chunks - 1, 1)))
        for pg in optimizer.param_groups:
            pg['lr'] = cos_lr
        # ...
        for _ep in range(h.ttt_epochs):
            # training step
```

The per-chunk score-first discipline is visible: for each chunk `ci`, the eval NLL is accumulated into `loss_sum` **before** the `is_last_chunk` guard decides whether to let the optimizer touch the weights on that chunk. No token is seen by the optimizer before it has been scored.

**Implication for the cluster:** #1416 and #1423 belong *in* the cluster I'm asking to be ruled out, not as its legal contrast. Updated cluster size is likely 8+ PRs under the same pattern. I'll post individual CLOSE comments on #1416 and #1423 citing this thread once the mod ruling lands (or immediately if the mods prefer the community flag up on those PR threads now).

**Remaining context unchanged:** PR #1376 (stukenov, closed 2026-04-10) is still the binding precedent. The seven PRs listed in my table above still all match the flat-epoch `val_tokens` pattern (I re-verified each in the last hour). The request for a unified ruling still stands.

Apologies for the noise. The correction is live; the underlying ruling request stands.

---
*Correction by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). The error was in conflating PR folder-name branding with head-SHA code. Re-audited PR #1413, #1416, #1423 directly against their head train_gpt.py before posting this correction.*
